web/pandas/pdeps/0013-standardize-apply.md

```diff
@@ -0,0 +1,333 @@
+
+Note also that `.pipe` operates on the `Series` while `apply`currently operates on each element in the data, so there is some differences that may have some consequence in some cases.
+
+This PDEP proposes that callables will be applies to the whole `Series`, so we in the future `Series.apply` will be as fast as `Series.pipe`.
+
+### 4. ufuncs in `Series.apply` vs. normal functions
+
+Performance-wise, ufuncs are fine in `Series.apply`, but non-ufunc functions are not:
+
+```python
+>>> %timeit large_ser.apply(np.sqrt)
+71.6 µs ± 1.17 µs per loop
+>>> %timeit large_ser.apply(lambda x:np.sqrt(x))
+63.6 ms ± 1.03 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
+```
+
+It is difficult to understand why ufuncs are fast in `apply`, while other callables are slow in `apply` (answer: it's because ufuncs operate on the whole Series, while other callables operate elementwise).
+
+This PDEP proposes that callables will be applies to the whole `Series`, so we in the future non-ufunc functions in `Series.apply` will be as fast as ufuncs.
+
+### 5. callables in `Series.apply` is slow, callables in `DataFrame.apply` is fast
+
+Above it was shown that using (non-ufunc) callables in `Series.apply` is bad performance-wise. OTOH using them in `DataFrame.apply` is fine:
+
+```python
+>>> %timeit large_ser.apply(lambda x: x + 1)
+24.3 ms ± 24 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
+>>> %timeit pd.DataFrame(large_ser).apply(lambda x: x + 1)
+160 µs +- 1.17 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
+```
+
+Having callables being fast to use in the `DataFrame.apply` method, but slow in `Series.apply` is confusing for users.
+
+This PDEP proposes that callables will be applies to the whole `Series`, so we in the future `Series.apply` will be as fast as `DataFrame.apply` already is.
+
+### 6. callables in `Series.apply` may fail, while callables in `DataFrame.apply` do not and vica versa
+
+```python
+>>> ser.apply(lambda x: x.sum())
+AttributeError: 'int' object has no attribute 'sum'
+>>> pd.DataFrame(ser).apply(lambda x: x.sum())
+0    6
+dtype: int64
+```
+
+Having callables fail when used in `Series.apply`, but work in `DataFrame.Apply` or vica versa is confusing for users.
+
+This PDEP proposes that callables will be applies to the whole `Series`, so callables given to `Series.apply` will work the same as when given to `DataFrame.apply`, so in the future we will have that:
+
+```python
+>>> ser.apply(lambda x: x.sum())
+6
+>>> pd.DataFrame(ser).apply(lambda x: x.sum())
+0    6
+dtype: int64
+```
+
+### 7.  `Series.apply` vs. `Series.agg`
+
+The doc string for `Series.agg` says about the method's `func` parameter: "If a function, must ... work when passed ... to Series.apply". But compare these:
+
+```python
+>>> small_ser.agg(np.sum)
+6
+>>> small_ser.apply(np.sum)
+0    1
+1    2
+2    3
+dtype: int64
+```
+
+Users would expect these two to give the same result.
```

Comment:
Ummm, not so sure about that.  What about the case of `np.abs()`  or just `abs` ?  In that case `Series.agg(abs)` gets applied element-wise!

Reply:
Those will give the same results, so the text is accurate?

I agree we cannot be guaranteed that `ser.agg(func)` will actually aggregate, but that is IMO a different discussion and maybe part of the discussion that has been opened about changing the API for this class of methods.